src/object/objecttype.c

```diff
@@ -668,14 +668,74 @@ exception_t decodeInvocation(word_t invLabel, word_t length,
             return EXCEPTION_SYSCALL_ERROR;
         }
 
-        setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
 #ifdef CONFIG_KERNEL_MCS
+
+        endpoint_t* ep_ptr = EP_PTR(cap_endpoint_cap_get_capEPPtr(cap));
+        if (endpoint_ptr_get_epThreshold(ep_ptr)!=0) {
+            
+            if(!canDonate) {
+                /* Only invocations that can donate are permitted to use a thresholded endpoint*/
+                current_syscall_error.type = seL4_IllegalOperation;
+                return EXCEPTION_SYSCALL_ERROR;
+            }
+
+            ticks_t required_budget; 
+            /* Perform addition, checking for overflow as we go */
+            if (getMaxTicksToUs() - NODE_STATE(ksConsumed) < endpoint_ptr_get_epThreshold(ep_ptr)) {
+                /* Overflow would occur */
+                required_budget = getMaxTicksToUs();
+            } else {
+                required_budget = endpoint_ptr_get_epThreshold(ep_ptr) + NODE_STATE(ksConsumed);
+            }
+
+            if (!available_budget_check(NODE_STATE(ksCurThread)->tcbSchedContext, required_budget)) {
+                
+                if (!block) {
+                    /* If we can't block, send fails silently, we don't wait for sufficient budget. */
+                    return EXCEPTION_NONE;
+                }
+
+                /* Charge the thread for its usage now 
+                * To get here, we passed through a MCS_DO_IF_BUDGET
+                * So the thread must have at least enough budget to cover ksConsumed
+                * So we can just commitTime().
+                */
+                assert(refill_sufficient(NODE_STATE(ksCurSC), NODE_STATE(ksConsumed)));
+                commitTime();
+
+                /* Yield and wait for sufficient budget */
+                if(!merge_until_budget(NODE_STATE(ksCurThread)->tcbSchedContext,endpoint_ptr_get_epThreshold(ep_ptr) + 2u * getKernelWcetTicks())) {
```

Comment:
Don't add WCET everywhere please. The threshold value should be twice WCET already.

Please write a test for sel4test that checks that the budget works as expected and use that to test your implementation. E.g. configure a task with a certain budget/period and check if IPC calls fail and succeed when they should.